File: Java/Combinations.java
/**
 * Problem Link: https://leetcode.com/problems/combinations/
 * 
 * Backtracking Problem, try out all the possible solutions and appropriate prunning 
 */
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        list = new ArrayList<List<Integer>>();
        combine(n, k, 1, new ArrayList<Integer>());
        return list;
    }
    
    private List<List<Integer>> list;
    
    private void combine(int n, int k, int start, ArrayList<Integer> ret){
        if(ret.size() == k) {
            list.add(new ArrayList<Integer>(ret));
            return;
        }
        if(ret.size() + n-start+1 < k) return;
        for(int i=start; i<=n; i++) {
            ret.add(i);
            combine(n, k, i+1, ret);
            ret.remove(ret.size()-1);
        }
    }
}

File: Swift/combinations.swift
/**
 * https://leetcode.com/problems/combinations/
 * 
 * 
 */ 
// Date: Thu Jun 18 11:14:48 PDT 2020
class Solution {
    func combine(_ n: Int, _ k: Int) -> [[Int]] {
        var solution: [[Int]] = []
        
        func choose(from nums: [Int], _ start: Int, _ left: Int, _ result: inout [Int]) {
            if left == 0 {
                solution.append(result)
                return
            }
            for index in start ..< nums.count {
                result.append(nums[index])
                choose(from: nums, index + 1, left - 1, &result)
                result.removeLast()
            }
        }
        
        var result: [Int] = []
        choose(from: Array((1...n)), 0, k, &result)
        return solution
    }
}
